File: scripts/process-books.py
"""
PDF Book Processor — extracts PDFs and seeds knowledge_chunks in Supabase
Usage: python3 scripts/process-books.py
Place PDFs in scripts/books/
"""
import os, re, sys, json, glob, urllib.request, urllib.error
from pathlib import Path

try:
    from pypdf import PdfReader
except ImportError:
    print("Installing pypdf..."); os.system("pip3 install pypdf --break-system-packages -q")
    from pypdf import PdfReader

# ── Load env ──────────────────────────────────────────────────────────────────
def load_env():
    env_path = Path(__file__).parent.parent / ".env.local"
    if not env_path.exists():
        print("ERROR: .env.local not found. Run: npx vercel env pull .env.local")
        sys.exit(1)
    for line in env_path.read_text().splitlines():
        if "=" in line and not line.startswith("#"):
            k, _, v = line.partition("=")
            os.environ.setdefault(k.strip(), v.strip())

# ── Supabase insert ───────────────────────────────────────────────────────────
def supabase_delete(url, key, source):
    """Delete existing chunks for this source (idempotent)"""
    req = urllib.request.Request(
        f"{url}/rest/v1/knowledge_chunks?source=eq.{urllib.parse.quote(source)}",
        method="DELETE",
        headers={"apikey": key, "Authorization": f"Bearer {key}"},
    )
    try: urllib.request.urlopen(req)
    except: pass

def supabase_insert(url, key, rows):
    import urllib.parse
    data = json.dumps(rows).encode()
    req  = urllib.request.Request(
        f"{url}/rest/v1/knowledge_chunks",
        data=data,
        method="POST",
        headers={
            "apikey": key,
            "Authorization": f"Bearer {key}",
            "Content-Type": "application/json",
            "Prefer": "return=minimal",
        },
    )
    try:
        urllib.request.urlopen(req)
        return True
    except urllib.error.HTTPError as e:
        print(f"  Insert error {e.code}: {e.read().decode()[:200]}")
        return False

# ── PDF processing ────────────────────────────────────────────────────────────
def extract_text(path):
    reader = PdfReader(path)
    parts  = []
    for page in reader.pages:
        text = page.extract_text() or ""
        parts.append(text)
    return "\n".join(parts)

def clean_text(text):
    text = re.sub(r'\s+', ' ', text)
    text = re.sub(r'[^\x20-\x7E\s]', '', text)
    return text.strip()

def chunk_text(text, source, chunk_size=400, overlap=60):
    words  = text.split()
    chunks = []
    i, n   = 0, 1
    while i < len(words):
        slice_   = words[i : i + chunk_size]
        content  = " ".join(slice_)
        if len(content) > 200 and '.' in content:
            preview = content[:80].rstrip()
            chunks.append({
                "title":    f"{source} — Part {n}: {preview}…",
                "source":   source,
                "category": "framework",
                "industry": None,
                "content":  content,
            })
            n += 1
        i += chunk_size - overlap
    return chunks

def process_pdf(path, source, supabase_url, supabase_key):
    print(f"\n📖  {source}")
    text    = clean_text(extract_text(path))
    words   = len(text.split())
    chunks  = chunk_text(text, source)
    print(f"    {words:,} words → {len(chunks)} chunks")

    # Delete old chunks for this book
    import urllib.parse
    req = urllib.request.Request(
        f"{supabase_url}/rest/v1/knowledge_chunks?source=eq.{urllib.parse.quote(source)}",
        method="DELETE",
        headers={"apikey": supabase_key, "Authorization": f"Bearer {supabase_key}"},
    )
    try: urllib.request.urlopen(req)
    except: pass

    # Insert in batches of 50
    batch_size = 50
    inserted   = 0
    for i in range(0, len(chunks), batch_size):
        batch = chunks[i : i + batch_size]
        if supabase_insert(supabase_url, supabase_key, batch):
            inserted += len(batch)
        print(f"    Inserted {inserted}/{len(chunks)}...", end="\r")

    print(f"    ✓ {inserted} chunks inserted      ")
    return inserted

def main():
    import urllib.parse
    load_env()

    supabase_url = os.environ.get("NEXT_PUBLIC_SUPABASE_URL", "").rstrip("/")
    supabase_key = os.environ.get("SUPABASE_SERVICE_ROLE_KEY", "")

    if not supabase_url or not supabase_key:
        print("ERROR: Missing NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY")
        sys.exit(1)

    books_dir = Path(__file__).parent / "books"
    pdfs      = sorted(books_dir.glob("*.pdf"))

    if not pdfs:
        print(f"No PDFs found in {books_dir}")
        sys.exit(0)

    print(f"Found {len(pdfs)} PDF(s)\n")
    total = 0
    for pdf in pdfs:
        source   = pdf.stem.replace("_", " ").replace("-", " ")
        inserted = process_pdf(str(pdf), source, supabase_url, supabase_key)
        total   += inserted

    print(f"\n✅  Done — {total} total chunks seeded into knowledge_chunks")

if __name__ == "__main__":
    main()

File: scripts/process-books.ts
/**
 * PDF Book Processor — converts PDFs in scripts/books/ into knowledge_chunks
 * Run: npx tsx scripts/process-books.ts
 */
import fs from "fs"
import path from "path"
import { PDFParse } from "pdf-parse"
import { createClient } from "@supabase/supabase-js"

let supabase: ReturnType<typeof createClient>

const BOOKS_DIR  = path.join(process.cwd(), "scripts/books")
const CHUNK_SIZE = 400   // words per chunk
const OVERLAP    = 60    // words overlap between chunks (preserves context across boundaries)

function cleanText(text: string): string {
  return text
    .replace(/\f/g, " ")                    // form feeds → space
    .replace(/\r\n/g, "\n")
    .replace(/[ \t]{3,}/g, "  ")           // collapse excessive spaces
    .replace(/\n{4,}/g, "\n\n\n")          // max 3 newlines
    .replace(/[^\x20-\x7E\n]/g, " ")       // strip non-ASCII garbage
    .trim()
}

function chunkText(text: string, source: string): Array<{ title: string; content: string; source: string }> {
  const words  = text.split(/\s+/).filter(w => w.length > 0)
  const chunks: Array<{ title: string; content: string; source: string }> = []
  let   i      = 0
  let   num    = 1

  while (i < words.length) {
    const slice   = words.slice(i, i + CHUNK_SIZE)
    const content = slice.join(" ")

    // Skip chunks that are mostly noise (very short or no real sentences)
    if (content.length > 150 && /[.!?]/.test(content)) {
      // Build a title from the first meaningful sentence fragment
      const firstLine = content.split(/[.!?\n]/)[0].trim().slice(0, 80)
      chunks.push({
        title:   `${source} — Part ${num}${firstLine ? ": " + firstLine : ""}`,
        content,
        source,
      })
      num++
    }

    i += CHUNK_SIZE - OVERLAP
  }

  return chunks
}

function inferCategory(filename: string): "framework" | "principle" {
  const lower = filename.toLowerCase()
  if (lower.includes("benchmark") || lower.includes("stat")) return "principle"
  return "framework"
}

async function processBook(filePath: string) {
  const filename  = path.basename(filePath, ".pdf")
  const source    = filename.replace(/_/g, " ").replace(/-/g, " ")
  const category  = inferCategory(filename)

  console.log(`\n📖 Processing: ${source}`)

  const buffer  = fs.readFileSync(filePath)
  const parser  = new PDFParse()
  const parsed  = await parser.parse(buffer)
  const cleaned = cleanText(parsed.text)

  console.log(`   ${parsed.numpages} pages · ${cleaned.split(/\s+/).length.toLocaleString()} words`)

  const chunks = chunkText(cleaned, source)
  console.log(`   → ${chunks.length} chunks`)

  // Delete existing chunks for this source (idempotent re-run)
  await supabase.from("knowledge_chunks").delete().eq("source", source)

  // Insert in batches of 50
  const BATCH = 50
  let inserted = 0
  for (let i = 0; i < chunks.length; i += BATCH) {
    const batch = chunks.slice(i, i + BATCH).map(c => ({
      title:    c.title,
      source:   c.source,
      category,
      industry: null,
      content:  c.content,
    }))
    const { error } = await supabase.from("knowledge_chunks").insert(batch)
    if (error) { console.error("   ✗ Insert error:", error.message); continue }
    inserted += batch.length
    process.stdout.write(`   Inserted ${inserted}/${chunks.length}...\r`)
  }

  console.log(`   ✓ Done — ${inserted} chunks inserted`)
}

async function main() {
  // Load env
  const envPath = path.join(process.cwd(), ".env.local")
  if (fs.existsSync(envPath)) {
    fs.readFileSync(envPath, "utf-8").split("\n").forEach(line => {
      const [k, ...v] = line.split("=")
      if (k && v.length) process.env[k.trim()] = v.join("=").trim()
    })
  }

  if (!process.env.NEXT_PUBLIC_SUPABASE_URL || !process.env.SUPABASE_SERVICE_ROLE_KEY) {
    console.error("Missing NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY in .env.local")
    process.exit(1)
  }

  supabase = createClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.SUPABASE_SERVICE_ROLE_KEY!
  )

  const files = fs.readdirSync(BOOKS_DIR)
    .filter(f => f.toLowerCase().endsWith(".pdf"))
    .map(f => path.join(BOOKS_DIR, f))

  if (!files.length) {
    console.log("No PDFs found in scripts/books/")
    process.exit(0)
  }

  console.log(`Found ${files.length} PDF(s) to process`)

  for (const file of files) {
    await processBook(file)
  }

  console.log("\n✅ All books processed. Knowledge base updated.")
}

main().catch(console.error)
